File: Tutorial5-2Dagger-Android-SubComponents/src/main/java/com/example/tutorial5_2dagger_android_subcomponents/di/AppModule.java
package com.example.tutorial5_2dagger_android_subcomponents.di;

import android.app.Activity;
import android.app.Application;
import android.content.Context;
import android.content.SharedPreferences;

import com.example.tutorial5_2dagger_android_subcomponents.MainActivity;

import javax.inject.Singleton;

import dagger.Binds;
import dagger.Module;
import dagger.Provides;
import dagger.android.ActivityKey;
import dagger.android.AndroidInjector;
import dagger.multibindings.IntoMap;

@Module(subcomponents = {ToastMakerSubComponent.class})
public abstract class AppModule {

    @Provides
    @Singleton
    static SharedPreferences provideSharedPreferences(Application application) {
        System.out.println("ApplicationModule context: " + application);
        return application.getSharedPreferences("PrefName", Context.MODE_PRIVATE);
    }


    /*
     *  NOTE: This is same as adding void inject(MainActivity mainActivity) to SubComponent as in
     *  non dagger-android, and can be removed using
     *  @ContributesAndroidInjector if SubComponent and it's Builder has NO methods
     */

    @Binds
    @IntoMap
    @ActivityKey(MainActivity.class)
    abstract AndroidInjector.Factory<? extends Activity>
    bindMainActivityFactory(ToastMakerSubComponent.Builder builder);
}

File: Tutorial5-2Dagger-Android-SubComponents/src/main/java/com/example/tutorial5_2dagger_android_subcomponents/di/AppModule.kt
package com.example.tutorial5_2dagger_android_subcomponents.di

import android.app.Application
import android.content.Context
import android.content.SharedPreferences
import com.example.tutorial5_2dagger_android_subcomponents.MainActivity
import dagger.Binds
import dagger.Module
import dagger.Provides
import dagger.android.AndroidInjector
import dagger.multibindings.ClassKey
import dagger.multibindings.IntoMap
import javax.inject.Singleton

@Module
abstract class AppModule {
    /*
     *  NOTE: This is same as adding void inject(MainActivity mainActivity) to SubComponent as in
     *  non dagger-android, and can be removed using
     *  @ContributesAndroidInjector if SubComponent and it's Builder has NO methods
     *
     * !! In kotlin Generic should match with @ClassKey, if it's ? then Kotlin should be *
     */
// 🔥 Injects ToastMakerSubComponent objects to MainActivity
    @Binds
    @IntoMap
    @ClassKey(MainActivity::class)
    abstract fun bindMainActivityFactory(builder: ToastMakerSubComponent.Builder): AndroidInjector.Factory<*>

    companion object {
        @JvmStatic
        @Provides
        @Singleton
        fun provideSharedPreferences(application: Application): SharedPreferences {
            println("ApplicationModule context: $application")
            return application.getSharedPreferences("PrefName", Context.MODE_PRIVATE)
        }
    }
}